### DISCUSSION TITLE
How do I turn off the logging?

### DISCUSSION THREAD

**Discussion by slhck** [mhenrixon/sidekiq-unique-jobs#608] [Q&A] [answered]
I'm using 7.0.4 and I'm getting messages like this printed all over the place:

```
2021-05-31T13:19:29.559Z pid=2247 tid=83r uniquejobs=client until_and_while_executing=uniquejobs:15508a50157d9050bf57bb4188e7a900 WARN: Timed out after 0s while waiting for primed token (digest: uniquejobs:15508a50157d9050bf57bb4188e7a900, job_id: 08b9921faed68072b76c299c)
```

The messages show when I am running Rspec tests and are directly printed to Rspec's STDOUT. They don't appear in the Rails logfile.

How do I turn these off?

I see that this behavior was changed from 6.x to improve debugging, but I don't need them at the moment.

My Sidekiq config is:

```rb
# frozen_string_literal: true

require 'sidekiq/api'
require 'prometheus_exporter/instrumentation'

Sidekiq.configure_server do |config|
  config.redis = { url: ENV['REDIS_URL'] }
  config.log_formatter = Sidekiq::Logger::Formatters::JSON.new
  config.on :startup do
    PrometheusExporter::Instrumentation::ActiveRecord.start(
      custom_labels: { type: "sidekiq" }, #optional params
      config_labels: [:database, :host] #optional params
    )
    PrometheusExporter::Instrumentation::SidekiqQueue.start
    PrometheusExporter::Instrumentation::Process.start type: 'sidekiq'
  end
  at_exit do
    PrometheusExporter::Client.default.stop(wait_timeout_seconds: 10)
  end
  config.server_middleware do |chain|
    chain.add PrometheusExporter::Instrumentation::Sidekiq
  end
  config.death_handlers << PrometheusExporter::Instrumentation::Sidekiq.death_handler

  config.client_middleware do |chain|
    chain.add SidekiqUniqueJobs::Middleware::Client
  end

  config.server_middleware do |chain|
    chain.add SidekiqUniqueJobs::Middleware::Server
  end
  SidekiqUniqueJobs::Server.configure(config)
end

Sidekiq.configure_client do |config|
  config.redis = { url: ENV['REDIS_URL'] }

  config.client_middleware do |chain|
    chain.add SidekiqUniqueJobs::Middleware::Client
  end
end

# All job classes that should repeat.
repeat_job_classes = [
    # TODO
]

# prevent dying jobs from littering uniqueness keys
Sidekiq.configure_server do |config|
  config.death_handlers << lambda { |job, _ex|
    if job['unique_digest']
      SidekiqUniqueJobs::Digests.del(digest: job['unique_digest'])
    end
  }
end

# Use this to clear out all existing jobs,
# and then start the repeating jobs directly.
ss = Sidekiq::ScheduledSet.new
repeat_job_classes.each do |rjc|
  ss.select { |retri| retri.klass == rjc.name }.each(&:delete)
  rjc.perform_async
end
```

My workers use:

```
sidekiq_options lock: :until_and_while_executing
```

**Comment by mhenrixon** [comment]
Better turn off unique jobs for testing. Is a section about that in readme.

**Reply by slhck** [reply]
I do need to test that unique jobs work, also during testing. Rspec was just an example.

In the main app, I get these messages on stdout as well, which is a bit less of an issue, but still, it would be better to have it configurable.

I was able to simply do:

```rb
SidekiqUniqueJobs.configure do |config|
  # disable verbose logging for testing
  config.logger = Logger.new("/dev/null") if Rails.env.test?
end
```

**Reply by mhenrixon** [reply]
It is unnecessary to test the gem. I mean, would you test that rails is doing what it should? 

I strongly recommend against it, just the way Mike Perham suggests against testing uniqueness for sidekiq enterprise. 

To turn off logging, just set the sidekiq unique loggers to a clone of the sidekiq logger or one that logs to dev null or turn the level up to fatal or something

**Reply by slhck** [reply]
> It is unnecessary to test the gem. I mean, would you test that rails is doing what it should?

In some edge cases, yes. Most of the business logic is explicitly expressed in the tests, so, there is one that says: "If you submit this job twice, it'll only be run once."

> I strongly recommend against it, just the way Mike Perham suggests against testing uniqueness for sidekiq enterprise.

I was going to read what you suggested, but the link in the README is broken, unfortunately.

> To turn off logging, just set the sidekiq unique loggers to a clone of the sidekiq logger or one that logs to dev null or turn the level up to fatal or something

Yes, the solution with `/dev/null` works just fine.

**Comment by mhenrixon** [comment, marked as answer]
https://github.com/mhenrixon/sidekiq-unique-jobs/discussions/608#discussioncomment-808133

```ruby
SidekiqUniqueJobs.configure do |config|
  # disable verbose logging for testing
  config.logger = Logger.new("/dev/null") if Rails.env.test?
end
```
